src/lattice_reaction_system_diffusion.jl

```diff
@@ -0,0 +1,461 @@
+    """The spatial reactions defined between individual nodes."""
+    spatial_reactions::Vector{<:AbstractSpatialReaction}
+    """The graph on which the lattice is defined."""
+    lattice::DiGraph
+
+    # Derrived values.
+    """The number of compartments."""
+    nC::Int64
+    """The number of edges."""
+    nE::Int64
+    """The number of species."""
+    nS::Int64
+    """Whenever the initial input was a di graph."""
+    init_digraph::Bool
+
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reactions::Vector{<:AbstractSpatialReaction},
+                                   lattice::DiGraph; init_digraph = true)
+        return new(rs, spatial_reactions, lattice, nv(lattice), ne(lattice),
+                   length(species(rs)), init_digraph)
+    end
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reactions::Vector{<:AbstractSpatialReaction},
+                                   lattice::SimpleGraph)
+        return LatticeReactionSystem(rs, spatial_reactions, graph_to_digraph(lattice);
+                                     init_digraph = false)
+    end
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reaction::AbstractSpatialReaction,
+                                   lattice::Graphs.AbstractGraph)
+        return LatticeReactionSystem(rs, [spatial_reaction], lattice)
+    end
+end
+# Covnerts a graph to a digraph (in a way where we know where the new edges are in teh edge vector).
+function graph_to_digraph(g1)
+    g2 = Graphs.SimpleDiGraphFromIterator(reshape(permutedims(hcat(collect(edges(g1)),
+                                                       reverse.(edges(g1)))), :, 1)[:])
+    add_vertices!(g2, nv(g1) - nv(g2))
+    return g2
+end
+# Gets the species of a lattice reaction system.
+species(lrs::LatticeReactionSystem) = species(lrs.rs)
+function diffusion_species(lrs::LatticeReactionSystem)
+    filter(s -> ModelingToolkit.getname(s) in getfield.(lrs.spatial_reactions, :species_sym),
+           species(lrs.rs))
+end
+
+# Gets the parameters in a lattice reaction system.
+function ModelingToolkit.parameters(lrs::LatticeReactionSystem)
+    unique(vcat(parameters(lrs.rs),
+                Symbolics.get_variables.(getfield.(lrs.spatial_reactions, :rate))...))
+end
+function compartment_parameters(lrs::LatticeReactionSystem)
+    filter(p -> !is_spatial_param(p, lrs), parameters(lrs))
+end
+function diffusion_parameters(lrs::LatticeReactionSystem)
+    filter(p -> is_spatial_param(p, lrs), parameters(lrs))
+end
+
+# Checks whenever a parameter is a spatial parameter or not. 
+function is_spatial_param(p, lrs)
+    hasmetadata(p, DiffusionParameter) && getmetadata(p, DiffusionParameter) &&
+        (return true)    # Wanted to just depend on metadata, but seems like we cannot implement that trivially.
+    return (any(isequal.(p, parameters(lrs.rs))) ? false : true)
```

Comment:
```suggestion
    return (any(isequal(p), parameters(lrs.rs)) ? false : true)
```
Again, try to avoid all this broadcasting -- it means you are allocating new memory when usually it is not needed.

Reply:
Also, what is the issue with depending on metadata?

Reply:
The problem is that (and this makes several things messy) when you call
```
DiffusionReaction(:D, :X)
```
it has no idea of what properties `D` was given in your `ReactionSystem`. So I don't want to append such metadata to `D` here. When the `LatticeReactionSystem` is created, we have all the information. However, metadata cannot be added to the parameters after they are created.

Reply:
That sounds like another reason to require `D` and `X` to be symbolics and not use `Symbols`. Seems like that would simplify things overall.

Reply:
I'm still not entirely sure we will get away from these problems, or at least it will require quite unsmooth work flows. Again, I cannot really remember more than having a decent amount of trouble figuring this out. I will look again and see what the best I can do is, and update here once I know,

Reply:
Similar to having a `@reaction k, A --> B` macro we could have a `@transport_reaction D X` macro (or something more compact if that is too long to write). Then we could have auto-discovery of diffusion parameters in the lattice system constructor from the symbolics within `DiffusionReaction`s, just as we do for normal `Reaction`s in the `ReactionSystem` constructor. But perhaps I am missing the fundamental issue.

Reply:
Never actually thought of broadcasting being a bad thing, that is good input.

Reply:
Fixed.